### ISSUE TITLE
Command: Modern Operations (1076160)

### ISSUE THREAD

**Issue by redpiller** [ValveSoftware/Proton#3807]
# Compatibility Report
- Name of the game with compatibility issues: Command: Modern Operations
- Steam AppID of the game: 1076160

<details><summary>System Information</summary><p>

```
Computer Information:
    Manufacturer:  Unknown
    Model:  Unknown
    Form Factor: Desktop
    No Touch Input Detected

Processor Information:
    CPU Vendor:  GenuineIntel
    CPU Brand:         Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz
    CPU Family:  0x6
    CPU Model:  0x2d
    CPU Stepping:  0x6
    CPU Type:  0x0
    Speed:  3300 Mhz
    16 logical processors
    8 physical processors
    HyperThreading:  Supported
    FCMOV:  Supported
    SSE2:  Supported
    SSE3:  Supported
    SSSE3:  Supported

    SSE4a:  Unsupported
    SSE41:  Supported
    SSE42:  Supported
    AES:  Supported
    AVX:  Supported
    AVX2:  Unsupported
    AVX512F:  Unsupported
    AVX512PF:  Unsupported
    AVX512ER:  Unsupported
    AVX512CD:  Unsupported
    AVX512VNNI:  Unsupported
    SHA:  Unsupported
    CMPXCHG16B:  Supported
    LAHF/SAHF:  Supported
    PrefetchW:  Unsupported

Operating System Version:
    Debian GNU/Linux 10 (buster) (64 bit)
    Kernel Name:  Linux
    Kernel Version:  4.19.0-8-amd64
    X Server Vendor:  The X.Org Foundation
    X Server Release:  12004000
    X Window Manager:  KWin
    Steam Runtime Version:  steam-runtime_0.20200318.2

Video Card:
    Driver:  NVIDIA Corporation GeForce GTX 760/PCIe/SSE2
    Driver Version:  4.6.0 NVIDIA 440.59
    OpenGL Version: 4.6
    Desktop Color Depth: 24 bits per pixel
    Monitor Refresh Rate: 60 Hz
    VendorID:  0x10de
    DeviceID:  0x1187
    Revision Not Detected
    Number of Monitors:  1
    Number of Logical Video Cards:  1
    Primary Display Resolution:  1920 x 1080
    Desktop Resolution: 1920 x 1080
    Primary Display Size: 23.62" x 13.39" (27.13" diag)
                                            60.0cm x 34.0cm (68.9cm diag)
    Primary Bus: PCI Express 16x
    Primary VRAM: 2048 MB
    Supported MSAA Modes:  2x 4x 8x 16x 
```

</p></details>

## I confirm:
- [x] that I haven't found an existing compatibility report for this game.
- [x] that I have checked whether there are updates for my system available.

## Symptoms 
Launcher starts, starting the game itself spawns a process then exits 

## Reproduction
Start the launcher, press the Start button, enjoy the silence

## Log File (Updated)
[steam-1076160.log.zip](https://github.com/ValveSoftware/Proton/files/4537785/steam-1076160.log.zip)

**Comment by kisak-valve** [comment]
Some lines of interest from the log:
`wine: Call from 0x7bc91c39 to unimplemented function MSVCR100_CLR0400.dll._initterm_e, aborting`

`wine: Call from 0x7bc91c39 to unimplemented function MSVCR100_CLR0400.dll._except_handler4_common, aborting`

**Comment by madewokherd** [comment]
Log is with native .NET, and MSVCR100_CLR0400.dll is native as well. Maybe a broken .NET install?

**Comment by redpiller** [comment]
I am so sorry, apparently i messed around with the prefix at some point in time, after deleting the application's compdata directory and letting it be recreated by proton, the behavior has changed (log updated)
sorry for wasting your time, please recheck the new log.

**Comment by kisak-valve** [comment]
Some more possible lines of interest:
```
[00000088:] EXCEPTION handling: System.InvalidProgramException: Invalid IL code in .: (): IL_0020: brfalse   IL_01a5

[00000088:] EXCEPTION handling: System.InvalidProgramException: Invalid IL code in .: (): IL_001e: callvirt  0x0a001185

...
Unhandled Exception:
System.InvalidProgramException: Invalid IL code in .: (string[]): IL_001d: callvirt  0x0a002011
```

**Comment by madewokherd** [comment]
Sounds like some kind of obfuscator, I'll see what I can find out.

**Comment by marthinwurer** [comment]
[steam-1076160.zip](https://github.com/ValveSoftware/Proton/files/5540328/steam-1076160.zip)

Tried to run it today, it crashed once I hit play on the launcher.

**Comment by marthinwurer** [comment]
Looked into it a bit more - the bytecode is obfuscated with SmartAssembly, which abuses bugs in the Microsoft .NET implementation which mono has a policy of not emulating. It looks like the solution would be to install Microsoft .NET, but that is failing with protontricks.

**Comment by mattipulkkinen** [comment]
The game still crashes as soon as you hit play in the launcher. I collected proton logs for this and mine only had one instance of "Invalid IL code", so maybe something has changed, or maybe not. I'm attaching the log anyway.

[steam-1076160.log](https://github.com/ValveSoftware/Proton/files/12308358/steam-1076160.log)

**Comment by SWarrener** [comment]
I testing this to see if there had been any progress. I was able to install .NET (version 4.5.2), however this didn't help with running the game. It now fails to even get to the launcher. The error appears to be that mscoree.dll is missing, but I checked and there are versions present in both the system32 and the syswow64 folders in the PFX.

[steam-1076160.log](https://github.com/ValveSoftware/Proton/files/12666303/steam-1076160.log)

**Comment by SWarrener** [comment]
Ignore the above comment from me, as that appears to be a broken prefix. 

However, I was able to make a small bit of progress. I installed dotnet472 (had to switch back to an earlier proton version (5.0-10) for this to work successfully) and specifically used windows 7 in the winetricks options dialogue that pops up. Switching back to current proton versions and then pressing play leads to a dotnet error message popping up, but then the launcher loading as before, but this time showing an ad on the left hand side. Pressing play leads to a black screen appearing after a few seconds. This hangs around for a couple more seconds and then closes. I've attached the log file below, but it appears to be a different error than the one in the above files.

[dotnet472-errorsplashprelaucnher-blackscreenpost-1076160.log](https://github.com/ValveSoftware/Proton/files/12675560/dotnet472-errorsplashprelaucnher-blackscreenpost-1076160.log)

**Comment by DrymarchonShaun** [comment]
Seeing the exact same thing with GE-Proton9-13 (none of the newer official versions would install dotnet), tried both dotnet472 and dotnet48 with no luck, most I got was the black box popping up for a second after hitting play in the launcher. Log is [here](https://gist.githubusercontent.com/DrymarchonShaun/0346876507303a0b0cae8f6dedd9341d/raw/31da4247641ce8ad6cdf12655d0f496e46dea018/steam-1076160.log)

EDIT - Got the game to launch with Proton-GE 9.13 and `dotnet48` + `d3dcompiler_47` installed through protontricks. Fonts are messed up, I plan on trying to fix them but it will have to wait till morning.

**Comment by DrymarchonShaun** [comment]
So I still need to double check the fonts used by the stable version because I think they changed fonts at some point but the public beta is using Microsoft's Segoe UI which they [refuse to let anyone redistribute](https://learn.microsoft.com/en-us/typography/fonts/font-faq)

> How about Segoe UI? I love it and would like to include that in my non-Windows app or game. Is it available from Monotype?
> 
> No, as Segoe UI is both our user interface and corporate branding font, it is not available for use outside of Microsoft products on non-Windows platforms.

**Comment by DrymarchonShaun** [comment]
Ok, so the stable version is working fine so far (besides the fonts looking terrible, but that's just because some of the fonts the game uses are only available on Windows, this is possible to fix by modifying the registry values under `Software\\Wine\\Fonts\\Replacements` in `pfx/user.reg`).

The beta version is *alright*, but there are some graphical bugs that I didn't see on stable, and the text in 'special message' popups as well as briefings is missing for some reason.

**Comment by SWarrener** [comment]
I also got the game working with DyrmarchonShaun's steps, although the fonts look fine for me.

For anyone finding this thread in the future, to get the game running use glorious eggroll 9.13 (newer versions may work, I haven't tested them) 

run the following command ```protontricks 1076160 -q dotnet48 d3dcompiler_47``` (or install the two components using the GUI)

The game should now run

**Comment by DrymarchonShaun** [comment]
IIRC font replacement is configured at the distro level, NixOS must not have it set up correctly/well. 
Tldr; Probably an issue on my end.

**Comment by SWarrener** [comment]
Just as we got it working, a new patch has somewhat set us back. The game renders HTML to display a bunch of critical player facing text, but currently none of that text is displayed (look at the briefing screen when loading a tutorial mission to see an example of this behaviour)

According to the official [patch notes](https://www.matrixgames.com/forums/viewtopic.php?f=10201&t=406696) the game has moved over to WebView2 to render HTML, but I haven't been able to get that to install properly through wine (various old forum posts have offered solutions but I haven't had any success with them)  

![NoHTMLRendering](https://github.com/user-attachments/assets/fd964f9d-e82e-4ad3-b835-a2ce9e940d0d)
If everything was working correctly there would be a bunch of text where the red writing is.

**Comment by DrymarchonShaun** [comment]
> According to the official [patch notes](https://www.matrixgames.com/forums/viewtopic.php?f=10201&t=406696) the game has moved over to WebView2 to render HTML, but I haven't been able to get that to install properly through wine (various old forum posts have offered solutions but I haven't had any success with them)

Ah, that explains what I was seeing in the beta version. I'll see if I can get it to work, *if downloading the update would stop locking up and crashing the steam client*

**Comment by SWarrener** [comment]
I was able to get an older version of Webview2 to install. I'm not allowed to post the link, but its version 109.0.1518.78, and a legitimate download can be trivially found online, just make sure you get the x64 version.  Newer versions don't install through wine.

After you've downloaded webview2, install it in the winepfx for the game. Once you've installed it, open winecfg (for the game's prefix), select mswebview2.exe in the list of applications and select windows 7 to run that executable. Apply the changes and exit. Everything should now work correctly

**Comment by kisak-valve** [comment]
Hello @SWarrener, the link you posted appears to be legally problematic and has been removed.

**Comment by SWarrener** [comment]
I think this is a legitimate source for the download (https://www.catalog.update.microsoft.com/Search.aspx?q=109.0.1518.78) as it comes from an official microsoft site. The download in question is the bottom option on the linked page.

**Comment by DrymarchonShaun** [comment]
> Everything should now work correctly

All I'm getting when I run either the x64 or x86 installer is a popup that says "unknown installer error."

**Comment by SWarrener** [comment]
> All I'm getting when I run either the x64 or x86 installer is a popup that says "unknown installer error."

Was it exactly the version I specified (109.0.1518.78)? I believe that is the only one that works. If it that version you were having problems with I'm afraid I don't have any suggestions other than waiting, as I think an option for webview2 may appear in winetricks/protontricks in the next release (whenever that is)

**Comment by DrymarchonShaun** [comment]
> Was it exactly the version I specified (109.0.1518.78)?

if you still have it, whats the `shasum` of the installer you used?
```shell

$ shasum microsoftedgestandaloneinstallerx64_402402f6bb4b801ffdd039705423f20dbb347cd6.exe 
402402f6bb4b801ffdd039705423f20dbb347cd6  microsoftedgestandaloneinstallerx64_402402f6bb4b801ffdd039705423f20dbb347cd6.exe
```

**Comment by SWarrener** [comment]
> > Was it exactly the version I specified (109.0.1518.78)?
> 
> if you still have it, whats the `shasum` of the installer you used?
> 
> ```shell
> $ shasum microsoftedgestandaloneinstallerx64_402402f6bb4b801ffdd039705423f20dbb347cd6.exe 
> 402402f6bb4b801ffdd039705423f20dbb347cd6  microsoftedgestandaloneinstallerx64_402402f6bb4b801ffdd039705423f20dbb347cd6.exe
> ```

shasum MicrosoftEdgeWebView2RuntimeInstallerX64.exe 
5743961fa14c36b107f87d922f432b02da9f7571  MicrosoftEdgeWebView2RuntimeInstallerX64.exe

This is what I have, although i used a different download link, which got deleted by the mods. Despite having the same version number it looks like the installer you used is different to the one I used. I can't post the download link I used on this thread, but if you were to take a look at this list (`Link removed by moderator`) and search for webview2....

**Comment by kisak-valve** [comment]
Hello @SWarrener, adding a layer of indirection doesn't make that any less legally problematic and your link has been removed.

**Comment by SWarrener** [comment]
> Hello @SWarrener, adding a layer of indirection doesn't make that any less legally problematic and your link has been removed.

Could you please specify how the link is legally problematic? Other linux compatibility tools seem to have no problem sharing the link, indeed the link I posted is associated with the bottles project. Also Microsoft themselves provide a free direct download link to what is meant to be exactly the same software

**Comment by DrymarchonShaun** [comment]
Ah, the one you used looks to be an older version of the 'Evergreen Standalone Installer' on this page - [https://developer.microsoft.com/en-us/microsoft-edge/webview2/?ch=1&form=MA13LH#download](https://developer.microsoft.com/en-us/microsoft-edge/webview2/?ch=1&form=MA13LH#download)

I'm not 100% sure what the difference between the file you had found and the one on the MS update catalog. I'll see if I can find an official link, and test to see if it does anything different.

**Comment by DrymarchonShaun** [comment]
So I'm not sure why but _any_ webview2 installers whether they install correctly or not corrupt the prefix or something and I get

```
$ protontricks 1076160 -q winecfg       
------------------------------------------------------
warning: You are using a 64-bit WINEPREFIX. Note that many verbs only install 32-bit versions of packages. If you encounter problems, please retest in a clean 32-bit WINEPREFIX before reporting a bug.
------------------------------------------------------
------------------------------------------------------
warning: You apppear to be using Wine's new wow64 mode. Note that this is EXPERIMENTAL and not yet fully supported. If reporting an issue, be sure to mention this.
------------------------------------------------------
Using winetricks 20240105 - sha256sum: 7e2686754b2c258a77ef5e525436f4e9d592b082c7c6a704815667b337725e78 with wine-9.0 (Staging) and WINEARCH=win64
installed_file1="Application could not be started, or no application associated with the specified file.
------------------------------------------------------
warning: bug: w_metadata art2kmin corrupt, might need forward slashes?
------------------------------------------------------
```
any time I try to run any protontricks commands. Not sure why

**Comment by SWarrener** [comment]
I've never seen anything like that before, have you tried on a completely fresh prefix? 

There are some people who have had various issues posting on the steam forums for the game, so that might be somewhere to look

**Comment by DrymarchonShaun** [comment]
Yeah, multiple times. I think it's a bug with protontricks, using the same `winetricks` it uses directly on the pfx folder manually works fine. Even after getting webview installed and setting it to windows 7 in `winecfg` I couldn't get it to work in game, but that might be caused by the weirdness I was seeing with protontricks.

**Comment by DrymarchonShaun** [comment]
So, following some instructions on protondb, I was able to get webview to install and function for some of the places its used, like the briefing you get when starting a mission. Opening the "side briefing" after loading the mission doesn't work though. Some UI elements are also turning black randomly, and I have to resize the window to get them to come back, before they disappear again.
